### PR TITLE
Screen Apps

### PR DESCRIPTION
**Pull request by StephGit**
Contains stories:

- [List applications and applicationServer](https://taiga.puzzle.ch/project/team-gurkensalat-board/us/885)
- [Create Applications](https://taiga.puzzle.ch/project/team-gurkensalat-board/us/887)
- [Create new appServer](https://taiga.puzzle.ch/project/team-gurkensalat-board/us/888)
- [Filter AppServers and Apps](https://taiga.puzzle.ch/project/team-gurkensalat-board/us/886)

### CODE REVIEW

File: AMW_angular/io/src/app/apps/app-add/app-add.component.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+@Component({
+  selector: 'app-app-add',
+  standalone: true,
+  imports: [FormsModule, NgSelectModule],
+  templateUrl: './app-add.component.html',
+})
+export class AppAddComponent {
+  @Input() releases: Signal<Release[]>;
+  @Input() appServerGroups: Signal<Resource[]>;
+  @Output() saveApp: EventEmitter<AppCreate> = new EventEmitter<AppCreate>();
```
Comment (mburri) [inline AMW_angular/io/src/app/apps/app-add/app-add.component.ts:19]: We could use `releases = input.required<Release[]>()` and `saveApp = output<AppCreate>()` instead of `@Input` and `@Output`

File: AMW_angular/io/src/app/apps/app-add/app-add.component.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+  constructor(public activeModal: NgbActiveModal) {
+    this.activeModal = activeModal;
+  }
+
+  hasInvalidGroup(): boolean {
+    const isInvalid =
+      this.appServerGroup === undefined || this.appServerGroup === null || this.appServerGroup?.releases.length === 0;
+    if (isInvalid) {
+      this.appServerRelease = undefined;
```
Comment (mburri) [inline AMW_angular/io/src/app/apps/app-add/app-add.component.ts:33]: this seems to be a side effect of this method. Why is this necessary?

File: AMW_angular/io/src/app/apps/app-create.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export interface AppCreate {
+  appName: string;
+  appReleaseId: number;
+  appServerId: number;
+  appServerReleaseId: number;
```
Comment (mburri) [inline AMW_angular/io/src/app/apps/app-create.ts:5]: Why are all properties prefixed with `app` - could the prefix be removed?

File: AMW_angular/io/src/app/apps/app-add/app-add.component.html [hunk 4]
```diff
@@ -0,0 +1,47 @@
+<div class="modal-header">
+  <h5 class="modal-title" id="editReleaseTitle">Create application</h5>
```
Comment (mburri) [inline AMW_angular/io/src/app/apps/app-add/app-add.component.html:2]: Why is the id set to `editReleaseTitle` - is the id used in any name?

File: AMW_angular/io/src/app/apps/app-server-add/app-server-add.component.html [hunk 5]
```diff
@@ -0,0 +1,26 @@
+<div class="modal-header">
+  <h5 class="modal-title" id="editReleaseTitle">Create application server</h5>
+  <button type="button" class="btn btn-light close" aria-label="Close" (click)="cancel()">
+    <span aria-hidden="true">&times;</span>
+  </button>
+</div>
```
Comment (mburri) [inline AMW_angular/io/src/app/apps/app-server-add/app-server-add.component.html:6]: Same as in `app-add.component.html` (and probably others too) - this is a candidate for a separete component.

File: AMW_angular/io/src/app/apps/apps-filter/apps-filter.component.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  standalone: true,
+  imports: [FormsModule, NgSelectModule],
+  templateUrl: './apps-filter.component.html',
+})
+export class AppsFilterComponent {
+  @Input() releases: Release[];
+
+  @Output() filterEvent = new EventEmitter<{ filter: string; releaseId: number }>();
+
+  releaseId: number = 50;
```
Comment (mburri) [inline AMW_angular/io/src/app/apps/apps-filter/apps-filter.component.ts:19]: What is the meaning of this default value? Why is it set to 50?

File: AMW_angular/io/src/app/apps/apps-list/apps-list.component.html [hunk 7]
```diff
@@ -0,0 +1,16 @@
+@if (apps && apps.length > 0) {
+<table class="table table-sm table-borderless mb-0">
+  <tbody>
+    @for (app of apps; track app) {
+    <tr [class.table-light]="!even">
```
Comment (mburri) [inline AMW_angular/io/src/app/apps/apps-list/apps-list.component.html:5]: I don't understand this. `even` is a property that gets passed into the component - how does it change when iterating over list of apps?

Comment (mburri) [reply AMW_angular/io/src/app/apps/apps-list/apps-list.component.html:5]: should probably be 
`@for (app of apps; track app; let even = $even) {`

and then remove to input property from the component

Comment (mburri) [reply AMW_angular/io/src/app/apps/apps-list/apps-list.component.html:5]: ouch - this is a nested table... then the input property makes sense.

But nesting tables seems to be a bad idea. Is this even valid html?

File: AMW_angular/io/src/app/apps/apps-servers-list/apps-servers-list.component.scss [hunk 8]
```diff
@@ -0,0 +1,6 @@
+table .w-10 {
+  width: 10%;
+}
+table .w-90 {
+  width: 90%;
+}
```
Comment (mburri) [inline AMW_angular/io/src/app/apps/apps-servers-list/apps-servers-list.component.scss:6]: duplicate of apps-list-component.scss

File: AMW_angular/io/src/app/apps/apps.component.ts [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+  canCreateApp = false;
+  canCreateAppServer = false;
+  canViewAppList = false;
+  isLoading = false;
+
+  currentPage: number;
+  lastPage: number;
+
+  loadingPermissions = computed(() => {
+    if (this.authService.restrictions().length > 0) {
```
Comment (mburri) [inline AMW_angular/io/src/app/apps/apps.component.ts:69]: refactor to a guard clause

File: AMW_angular/io/src/app/apps/apps.component.ts [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    this.isLoading = false;
+  }
+
+  private setPagination() {
+    this.currentPage = Math.floor(this.offset() / this.maxResults()) + 1;
+    this.lastPage = Math.ceil(this.count() / this.maxResults());
+  }
+
+  private getUserPermissions() {
+    this.canCreateApp = this.authService.hasResourcePermission('RESOURCE', 'CREATE', 'APPLICATION');
```
Comment (mburri) [inline AMW_angular/io/src/app/apps/apps.component.ts:91]: Check: does this make a backend request for each permission?

File: AMW_angular/io/src/app/apps/apps.component.ts [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+    this.isLoading = true;
+    this.appsService
+      .createAppServer(appServer)
+      .pipe(takeUntil(this.destroy$))
+      .subscribe({
+        next: () => {
+          this.toastService.success('AppServer saved successfully.');
+        },
+        error: (e) => {
+          this.error$.next(e.toString());
```
Comment (mburri) [inline AMW_angular/io/src/app/apps/apps.component.ts:121]: does `this.isLoading` still apply for the error case?

Comment (mburri) [reply AMW_angular/io/src/app/apps/apps.component.ts:121]: should we introduce a remote data pattern?

File: AMW_angular/io/src/app/apps/apps.service.ts [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        })
+        .pipe(catchError(this.handleError));
+    }
+
+    return this.http
+      .post<AppCreate>(`${this.appsUrl}?appName=${app.appName}&releaseId=${app.appReleaseId}`, app, {
+        headers: this.getHeaders(),
+      })
+      .pipe(catchError(this.handleError));
+  }
```
Comment (mburri) [inline AMW_angular/io/src/app/apps/apps.service.ts:87]: candidate to extract two methods that help to understand the difference of the two REST calls.

File: AMW_angular/pom.xml [hunk 13]
```diff
@@ -57,23 +57,23 @@
                         </goals>
                     </execution>
                     <execution>
-                        <id>npm ci</id>
+                        <id>npm install</id>
```
Comment (mburri) [inline AMW_angular/pom.xml:60]: Why? Is there a reason to change it from npm ci to npm install? What is the difference between the two of them?

Comment (mburri) [reply AMW_angular/pom.xml:60]: we should use `npm ci` here and not `npm install`

See: https://stackoverflow.com/a/53325242

File: AMW_business/src/main/java/ch/puzzle/itc/mobiliar/business/apps/boundary/ListAppsUseCase.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+
+import ch.puzzle.itc.mobiliar.business.resourcegroup.entity.ResourceWithRelations;
+import ch.puzzle.itc.mobiliar.common.exception.NotFoundException;
+import ch.puzzle.itc.mobiliar.common.util.Tuple;
+
+import java.util.List;
+
+public interface ListAppsUseCase {
+
+    Tuple<List<ResourceWithRelations>, Long> appsFor(Integer startIndex, Integer maxResults, String filter, Integer releaseId) throws NotFoundException;
```
Comment (mburri) [inline AMW_business/src/main/java/ch/puzzle/itc/mobiliar/business/apps/boundary/ListAppsUseCase.java:11]: We coud create a dedicated object that helps to understand what the Long value stands for - instead of using a Tuple here.

File: AMW_business/src/main/java/ch/puzzle/itc/mobiliar/business/apps/control/AppsService.java [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+import ch.puzzle.itc.mobiliar.common.util.Tuple;
+
+import javax.ejb.Stateless;
+import javax.inject.Inject;
+import java.util.List;
+
+import static ch.puzzle.itc.mobiliar.business.security.entity.Action.CREATE;
+
+@Stateless
+public class AppsService implements ListAppsUseCase, AddAppServerUseCase, AddAppUseCase, AddAppWithServerUseCase {
```
Comment (mburri) [inline AMW_business/src/main/java/ch/puzzle/itc/mobiliar/business/apps/control/AppsService.java:31]: I'm not sure if I really like it that this Service implements all use cases. Maybe its ok. I have to think about it...

File: AMW_rest/src/main/java/ch/mobi/itc/mobiliar/rest/dtos/AppAppServerDTO.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+import javax.xml.bind.annotation.XmlRootElement;
+
+
+
+@XmlRootElement(name = "appAppServer")
+@XmlAccessorType(XmlAccessType.FIELD)
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class AppAppServerDTO {
```
Comment (mburri) [inline AMW_rest/src/main/java/ch/mobi/itc/mobiliar/rest/dtos/AppAppServerDTO.java:20]: naming

File: AMW_rest/src/main/java/ch/mobi/itc/mobiliar/rest/dtos/AppAppServerDTO.java [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+    String appName;
+    Integer appReleaseId;
+    Integer appServerId;
+    Integer appServerReleaseId;
```
Comment (mburri) [inline AMW_rest/src/main/java/ch/mobi/itc/mobiliar/rest/dtos/AppAppServerDTO.java:24]: redundant prefix

**Review by mburri — CHANGES_REQUESTED** [review]
I've added some comments - mostly for me instead of creating tasks.

The behaviour changes are correct as far as I know. But it would have been nice to split the PR in smaller chunks.

File: AMW_angular/io/src/app/apps/app-add/app-add.component.ts (hunk 2 above)
Comment (StephGit) [reply AMW_angular/io/src/app/apps/app-add/app-add.component.ts:33]: This resets the ng-select for appServerReleases in case the appServer has been reset/unselected

File: AMW_angular/io/src/app/apps/apps.component.ts (hunk 10 above)
Comment (StephGit) [reply AMW_angular/io/src/app/apps/apps.component.ts:91]: No, there is one call for all user permissions which are provided by the service

File: AMW_angular/pom.xml (hunk 13 above)
Comment (StephGit) [reply AMW_angular/pom.xml:60]: Sorry, changed to pnpm in first position, there is no pnpm ci ....npm ci is correct

File: AMW_angular/io/src/app/apps/apps.component.ts (hunk 9 above)
Comment (StephGit) [reply AMW_angular/io/src/app/apps/apps.component.ts:69]: Not sure, this is needed to receive the signal update after userpermissions are provided

File: AMW_angular/io/src/app/apps/app-add/app-add.component.html (hunk 4 above)
Comment (StephGit) [reply AMW_angular/io/src/app/apps/app-add/app-add.component.html:2]: Copy paste error

**Review by mburri — APPROVED** [review]
